---
title: "An Introduction to celda"
author: Josh D. Campbell, Sean Corbett, Masanao Yajima, Zichun Liu, Shiyi Yang, Tianwen
  Huan, Anahita Bahri
date: '`r Sys.Date()`'
output:
  pdf_document: default
  html_document: default
vignette: |
  %\VignetteIndexEntry{Vignette Title} %\VignetteEngine{knitr::rmarkdown} \usepackage[utf8]{inputenc}
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,fig.width=8,fig.height=7 ,
			out.width="0.7\\linewidth", fig.align = "center")
library(RColorBrewer)
library(ggplot2)
library(celda)
library(gtools)
library(pheatmap)
library(reshape2)
library(vcd)
library(corrplot)
library(gtable)
library(scales)


```

# 1 Introduction

The purpose of this vignette is to briefly introduce package `celda`. `celda` is the abbreviation of "CELlular Latent Dirichlet Allocation."

It means that we implement Bayesian Hierarchical Modeling for Clustering Single-Cell RNA-Seq Data

To install `celda`:  

```{r,eval=FALSE}
library(devtools)
install_github("definitelysean/celda")
```

The LDA model in `celda` package can be visualized as follows:

\includegraphics[height=7cm]{graph.dot.png}


## 2.1 An Example with built-in Toy dataset 

### 2.1.1 Attach dataset
```{r}
library(celda)
data("sample.cells")
```

### 2.1.2 Initial Visualization

```{r}
sample.cells
```

```{r}
pheatmap::pheatmap(scale(sample.cells,center = F),cluster_rows = FALSE, cluster_cols = FALSE)
#pheatmap::pheatmap(log(sample.cells + 1),cluster_rows = FALSE, cluster_cols = FALSE)

```

Our sample have 4 types of cell and 4 types of transcripte states.

### 2.1.3 Run model with correct K and L 

Cluster cell on sample:

```{r}
toy_celda_c = celda_C(sample.cells,K = 4, sample.label = rep(c(1,2),c(6,6)))
```

Cluster both cell and gene: 

```{r}
toy_celda_cg = celda_CG(counts = sample.cells,sample.label = rep(1,12),K =4, L =4 )
```

### 2.1.4 Plot heatmap with celda cell/gene labels
```{r}
#celda_heatmap(counts = sample.cells, K = 4, z = toy_celda_c$z, L = 2, y = rep(c(1,2),c(5,5)))


```


## 2.2 An example using simulation data
With the cursory background out of the way, weâ€™ll proceed with some examples to help understand how the package is used.


### 2.2.1 Generate data using built-in function

celda provide a built-in method to generate simulation data. Directly call `cCG.generateCells`. We can specify some arguments in the function:

  S: Number of Samples
 
  C.Range: Range of number of cells in a sample 
 
  N.Range: Range of number of transcripts in a cell
 
  G: Total number of Gene types
 
  K: Total number of cell populations
 
  L: Total number of transcriptional states
 
  alpha: Assuming the type of cell(K types) in S samples follows a multinomial distribution. $\alpha$ is the parameter of its conjugate priors distribution 
 
  beta: Assuming the type of transcriptional states(L types) in K type of cell follows a multinomial distribution. $\beta$ is the parameter of its conjugate priors distribution that generate that multinomial distribution.
 
  gamma: Assuming the distribution of a gene contribute to a transcriptional states(L types) is a multinomial distribution. $\gamma$ is the parameter of its conjugate priors distribution that generate that multinomial distribution.
 
  delta: Assuming the proportion of different types of gene in a transcriptional state  follows a multinomial distribution. $\delta$ is the parameter of its conjugate priors distribution that generate that multinomial distribution. 
 
  seed: Random seed

```{r simulate data}
sim_counts = simulateCells.celdaCG(G = 100,K = 3, L = 5)
str(sim_counts)
```


The `cCG.generateCells` will return a list with standard format of celda. In the output:
  
  z: Indicate which kind of cell it is for all cells in all samples
 
  y: Indicate which transcriptional state a gene contributed to 
 
  sample: Indicate which sample a cell is from
 
  counts: In one cell what kind of gene and how many times it worked. (gene -> transcripts -> cell)
 
  theta: Proportion(multinomial distribution) of different types of cells in samples
 
  phi: Proportion(multinomial distribution) of different types of transcriptional states in cells
 
  psi: Proportion(multinomial distribution) of different types of genes in transcriptional states
 
  eta: Probability(multinomial distribution) of a gene contribute to different types of transcriptional states


### 2.2.2 LDA Clustering on Cells and Genes

We cluster cells and genes using `celda_CG`, here we want to cluster cells into 3 categories and cluster genes into 5 categories. And due to time consumption, we limit the iteration to be 25.

K = Total number of cell populations
L = Total number of transcriptional states

```{r}
celda_cluster_assignments = celda_CG(sim_counts$counts, sample.label = sim_counts$sample,
                                     K = 3, L = 5, max.iter = 25)
str(celda_cluster_assignments)
```

z: Cell cluster assignments from the final iteration of Gibbs sampling
 
y: gene cluster assignments from the final iteration of Gibbs sampling 

complete.z: History of cell cluster assignments across all iterations of Gibbs

complete.y: History of gene cluster assignments across all iterations of Gibbs

z.stability: [0,1] measure of stability for the cell clustering chain

y.stability: [0,1] measure of stability for the gene clustering chain

complete.z.stability: History of z.stability over all iterations of Gibbs sampling

complete.y.stability: Historyof y.stability over all iterations of Gibbs sampling

completeLogLik: Log-likelihood of all gene and cell cluster assignments over all iterations of Gibbs sampling

finalLogLik: Log-likelihood of final gene and cell cluster assignments

z.prob:

y.prob:

### 2.2.4 Cluster results and visualization

Because we generate the data using LDA model, we can know how well our cluster result is.

```{r}
xtabs(~celda_cluster_assignments$z+ sim_counts$z)

xtabs(~celda_cluster_assignments$y + sim_counts$y)

```

```{r }
#### Heatmap, Mosaic plot and corrplot on Cluster cell data
samp = sim_counts$sample

cluster_z = celda_cluster_assignments$z

tb = xtabs(~samp + cluster_z)

nmat <- tb/rowSums(tb)

mnmat<-melt(nmat)

mnmat$samp<-factor(mnmat$samp)

mnmat$cluster_z<-factor(mnmat$cluster_z,levels=sort(unique(mnmat$cluster_z),TRUE))
```

```{r fig.width=5,fig.height=3 }
#ggplot(mnmat)+geom_tile()+
#  aes(x=samp,y=cluster_z,fill=value)+scale_fill_gradient2()+
#  scale_x_discrete(position = "top")+xlab("sample")+ylab("cluster")
```

```{r}
corrplot(t(nmat),method = "circle",is.corr = FALSE)
```


```{r}
#### Use celda_heatmap and semi_pheatmap to produce heatmap
#celda_heatmap(sim_counts$counts,
#              K = 3,z = celda_cluster_assignments$z,
#              L = 5,y = celda_cluster_assignments$y)
```

```{r}
#celda_heatmap(log(sim_counts$counts+1),
#              K=3,z = celda_cluster_assignments$z, 
#              L = 5,y = celda_cluster_assignments$y)
```

```{r}
#celda_heatmap(t(apply(sim_counts$counts,1,scale)),
#              K=3,z = celda_cluster_assignments$z, 
#              L = 5,y = celda_cluster_assignments$y)
```

```{r}
semi_pheatmap(t(apply(sim_counts$counts,1,scale)),
              row_label = paste("gene",1:dim(sim_counts$counts)[1],sep = ""),
              col_label = paste("cell",1:dim(sim_counts$counts)[2],sep = ""),
              cutree_cols = 3,
              cutree_rows = 5)
```


# 3 Selecting the number of clusters  (and other diagnostic plots)

## 3.1 Run the toy dataset with celda wrapper with multiple K/L choices
```{r}

```

## 3.2 Generate plots for K and L selection
```{r}

```

# 4 Description of different models using toy dataset
## 4.1 Cells (C)
```{}
```

## 4.2 Genes (G)
```{}
```

## 4.3 Cells into populations and Genes into transcriptional states (CG)
```{}
```

## 4.4 Genes into Mixtures of transcriptional states (M)
```{}
```

## 4.5 Cells into populations and genes into Mixtures of transcriptional states (CM)
```{}
```


# 5 Real dataset example (10X)

## 5.1 Read in data
```{r}

```

## 5.2 Filter genes
```{r}

```

## 5.3 Run celda wrapper
```{r}

```

## 5.4 Select parameters
```{r}

```

## 5.5 Plot heatmap
```{r}

```


# Appendix
  
```{r,echo=FALSE,eval=FALSE}
modgraph<-'
Digraph fig1 {
rankdir = LR; //order things from left to right

//define alpha and beta as existing
alpha [shape=circle,label=<&alpha;>];
beta [shape=circle,label=<&beta;>];
gamma [shape=circle,label=<&gamma;>];
//not strictly nescessary but helps if you want to
//assign them specific shapes or colours

subgraph cluster_S
{
    label = "S"
    psi [shape=circle,label=<&psi;>];
  psi ->y;

subgraph cluster_M 
{
    label = <M<sub><font point-size="8">s</font></sub>>

 
    y [shape=circle];
    subgraph cluster_N
    {
        label = <N<sub><font point-size="8">ms</font></sub>>
        z [shape=circle];
        x [shape=circle, style=filled]
        z->x; 
    }
  y->z;

}
}
alpha -> theta;
beta -> phi;
gamma ->psi;
  subgraph cluster_L 
  {
      label ="L"
      theta [shape=circle,label=<&theta;>];
    theta -> z;
  }
  subgraph cluster_K
  {
      label = "K"
      phi [shape=circle,label=<&phi;>];
      phi->x;
  }
}

'
con <- file("graph.dot", "w")
cat(modgraph, file=con)
close(con)
#system("dot -Txdot graph.dot| dot2tex > graph.tex")
system("dot -T png -O graph.dot")
```

